Add get(coll, key) for Associative returning Nullable

dbeach24 opened the pull request:
This returns a Nullable value corresponding to the key (or null).
Includes specialization for dictionaries.
Fixes #13055

dbeach24 committed:
Added implementation of get(coll, key) -> Nullable.
This returns a Nullable value corresponding to the key (or null).
Includes specialization for various dictionary types.
(Fix bug in priority queue test.)
Fixes #13055

tkelman commented on base/dict.jl:
```diff
@@ -6,6 +6,13 @@ const secret_table_token = :__c782dbf1cf4d6a2e5e3865d7e95634f2e09b5902__
 haskey(d::Associative, k) = in(k,keys(d))
 
+"""
+    getnull(d::Associative, key)
```
if exported, this should also go in the rst docs - see CONTRIBUTING.md for full explanation, but you put the signature there then run `make docs` and the docstring content should get populated over.

tkelman commented on base/dict.jl:
```diff
@@ -6,6 +6,13 @@ const secret_table_token = :__c782dbf1cf4d6a2e5e3865d7e95634f2e09b5902__
+"""
+    getnull(d::Associative, key)
+
+Return Nullable(`value`) corresponding to `key` if present, otherwise null.
```
backticks around `Nullable(value)` would make more sense then around `value` ?

vtjnash commented:
would the names `tryget` or `trygetindex` be better than `getnull` (e.g. like `tryparse`)?

vtjnash commented:
or even just define this is 2-argument `get`?

TotalVerb commented:
This should get its own name. Otherwise, it will not generalize to multidimensional indexed arrays.

dbeach24 commented:
@vtjnash @TotalVerb Based on your comments, I changed the name to `tryget(coll, key)`.
@tkelman I have updated the docs as you suggested.

Please let me know if you find any other issues.

tkelman commented on base/dict.jl:
```diff
@@ -698,6 +706,11 @@ function get{K,V}(default::Callable, h::Dict{K,V}, key)
 
+function tryget{K,V}(h::Dict{K,V}, key)
+    index = ht_keyindex(h, key)
+    return (index<0) ? Nullable{V}() : Nullable{V}(h.vals[index]::V)
```
is this more efficient than the abstract definition above? the generic one should be correct here right?

TotalVerb replied:
This avoids computing the hash twice.

dbeach24 replied:
Yes, it avoids computing the hash and looking up the slot in the table twice.
@tkelman Please see the original issue #13055.

tkelman commented on base/dict.jl:
```diff
@@ -836,6 +850,14 @@ function get(dict::ImmutableDict, key, default)
     return default
 end
 
+function tryget{K,V}(dict::ImmutableDict{K,V}, key)
```
not currently covered by tests?

dbeach24 replied:
Yes, it is covered by tests.

dbeach24 commented on test/dict.jl:
```diff
@@ -452,6 +456,10 @@ let d = ImmutableDict{String, String}(),
     @test get(d4, "key1", :default) === v2
     @test get(d4, "foo", :default) === :default
     @test get(d, k1, :default) === :default
+    @test tryget(d1, "key1") === Nullable{String}(v1)
```
`tryget(::ImmutableDict)` is tested here.

eschnett commented:
The name `tryget` sounds as if this function was connected to a `try` statement. I'd rather have something like `nullable` in the name, as in `get_nullable`. This function is unlikely to be used too frequently, and making its action explicit in its name is a good thing.

dbeach24 commented:
@eschnett `tryget` was chosen because it is consistent with the existing `tryparse` (which also returns a `Nullable`).

As for the function being unlikely to be used, I will point out that in Python, for example, `dict.get(key)` returns `None` if the key is not present, and this is not considered an esoteric use of dictionaries.  Of course, Python doesn't have to worry about type stability, so it is simply the default case of `dict.get(key, default=None)`.

Julia seems to lack a type-stable equivalent, which is what motivated this addition.

eschnett commented:
I understand and agree with the motivation. What I meant to say is that the difference in function name length between `tryget` and `get_nullable` is here not that important because most codes will not contain tens of calls to `get_nullable` in a single function, and `get_nullable` is easier to understand.

quinnj commented:
FWIW, I'm not a fan of `tryget` either, mainly because I don't like`tryparse` either. Eventually, I think we should do `parse(Nullable{Int}, x)` if you want a Nullable back instead of needing a separate method.

I wonder if it'd be worth considering an alternative method like:

```julia
julia> Base.get{T}(::Type{T}, dict, key, default) = haskey(dict, key) ? T(dict[key]) : T(default)

julia> t = Dict{String,Int}("a"=>1, "b"=>2, "c"=>3)
Dict{String,Int64} with 3 entries:
  "c" => 3
  "b" => 2
  "a" => 1

julia> get(Nullable{Float64}, t, "a", 0)
Nullable{Float64}(1.0)

julia> get(Nullable{Float64}, t, "d", 0)
Nullable{Float64}(0.0)

julia> get(Int8, t, "a", 0)
1

julia> get(Int8, t, "d", 0)
0
```

dbeach24 commented:
@eschnett @quinnj I'd like to narrow the scope of this PR.  Can we please move debates about renaming the existing `try...` functions to a different issue?

I'm trying to resolve the narrow issue of avoiding repeat lookups in a dictionary when both testing for a key and retrieving its value.  When @vtjnash pointed out the existing `tryparse` function, I tried to be consistent with that in choosing the name `tryget`.

If the community later concludes that all the `try...` functions are bad, and, for example, that `get` with four arguments is better, then someone can come along and implement those changes in a separate PR.

StefanKarpinski commented:
I think the name `tryget` is more consistent with what we already have. If we come up with a better scheme for expressing this kind of thing (which I think we should), then maybe we can change it, but for now, this name seems as good as any other.

quinnj commented:
I still think 4-arg `get` is worth considering, even at this point, since it avoids introducing a completely new method to Base in addition to being more functional than the currently proposed `tryget`.

dbeach24 commented:
If we must go there, then I'll say that:

```julia
x = get(Nullable{Int64}, dict, "myKey", Nullable{Int64}())
```

makes my eyes bleed.  More general or not, I would want to have a succinct and readable way of expressing this.  Spelling out the Nullable-decorated value type of the dictionary (twice!) is something only a Java programmer could love.  (duck!)

```julia
x = tryget(dict, "myKey")
```

(or `getnull` or even `get_nullable`) seems far more readable to me.

StefanKarpinski commented:
Maybe just `get(Nullable, dict, "key")`?

dbeach24 commented:
@StefanKarpinski Is there any precedent in the language for doing something like that?  Admittedly, it's no longer than `get_nullable(dict, "key")`...  What would the definition look like?

martinholters commented:
> Maybe just `get(Nullable, dict, "key")`?

Nice and concise. Combined with @quinnj's proposal, it would just be something like
```Julia
get(::Type{Nullable}, dict, key) = get(Nullable{valtype(dict)}, dict, key)
function get{T}(::Type{Nullable{T}}, dict, key, default=Nullable{T}())
# ...
end
```
I guess?

It only seems the semantics would be ambiguous if someone decides to store `Nullable`s in a dict.

TotalVerb commented:
It would be less ambiguous (and more general) if it were `get(f, dict, i)`. The neutral element could perhaps be determined through multiple dispatch.

dbeach24 commented:
@TotalVerb So we would presumably define:

```julia
f() = Nullable{Int}()    # neutral element
f(x) = Nullable{Int}(x) # present element
```

And `get` would call the first form if the element is not present...
So, in this instance, `f` could actually be `Nullable{Int}`, as in:

```julia
get(Nullable{Int}, dict, key)
```

With the function coming first, this almost seems to suggest a `do` construct, but I can't see how that's useful in this instance:

```julia
x = get(dict, "key) do
   # ???
end
```

TotalVerb commented:
Unfortunately, I just realized that three-argument `get` is already defined and taken, and the semantics are not the same. So that syntax is off the table.

dbeach24 commented:
@TotalVerb Doesn't the existing 3-arg `get` have the associative container as the first arg?  If our has the associative as the second arg, can we define both forms without ambiguity?

```julia
get(x::Associative, key, default) # use this signature for existing 
get(f::Function, x::Associative, key) # use this signature for the new proposed form
```

Maybe this is hard to get right -- I don't know.

TotalVerb commented:
There is also a 3-arg get with function as first argument. It tries to get the value, then returns it unchanged; otherwise, it returns the value obtained by invoking the function.

dbeach24 commented:
@TotalVerb Amazingly (I think) this function is almost exactly what we want, but we'd have to change the semantics.  The current behavior is:

```julia
get(f, dict, key) = haskey(dict, key) ? dict[key] : f()
```

The behavior we want is:

```julia
get(f, dict, key) = haskey(dict, key) ? f(dict[key]) : f()
```

(Caveat: The real implementation would need to lookup the key only once, not twice as shown here.)

It's close, but I guess this would constitute a breaking change if implemented.

TotalVerb commented:
The trouble is that changing these semantics is a pretty big breaking change, especially since I believe this form is used often with `do` blocks:

```julia
get(dict, key) do
    # create default
end
```

dbeach24 commented:
Perhaps the cure is worse than the disease.

What guiding principles does the Julia community use to determine when it's better to overload an existing method, and when it's better to just have a separately named one?

Personally, I'm starting to lean toward the latter again.

TotalVerb commented:
I honestly am really not a fan of using the same name for what are very, very different operations. We don't use the same name for `getindex` and `get`, and I don't understand why this operation needs the same name.

rfourquet commented:
@TotalVerb I don't understand why this could not be the 2-arg `get`, "[because] it will not generalize to multidimensional indexed arrays", as `get` is not defined for arrays... Would you mind clarifying this point?

TotalVerb commented:
@rfourquet I am going to walk back my argument, because in fact `get` as it exists is fundamentally non-generic and cannot be made general to n-dimensional arrays, on account of its argument order. Functions like `setindex!` put the indices last so that they can work on arrays; `get` is not defined this way and so it could not possibly apply to arrays.

dbeach24 commented:
In light of all of this, I'd like to propose that we use the 2-argument form of `get` to return a nullable value from any associative.

```julia
get{K,V}(a::Associative{K,V}, key) = haskey(a, key) ? Nullable{V}(a[key]::V) : Nullable{V}()
```

This would seem to have the following benefits:

1. Does not introduce any new exports from Base.
2. Does not conflict with existing definitions of get.
3. Type-stable.
4. Succinct.
5. Can be specialized to avoid double-lookup for various dictionary implementations.
6. Semantically consistent with Python's `dict.get(key, default=None)`.

Up-Vote / Down-Vote. Reactions?

quinnj commented:
I can get behind that.

Though I would note that this would seem to be our first big foray into a real "commitment" to Nullables. Not saying it's a bad thing or irreversible or anything like that, I just know it took a while for Nullables to come about and this is a big (probably good) step to really integrating them with the rest of Base.

dbeach24 commented:
@quinnj That's a very good point.  I hope that committing to `Nullable` is a good thing, but maybe that's not a given.

I remember hearing at least one opinion at this year's JuliaCon that `Nullable` was a less-than-elegant solution ported over from more rigid, statically-typed languages.  It was suggested that if Julia's compiler were enhanced to emit optimized dispatch code, particularly in the case where the possible types of a variable are fairly constrained, that returning `Union{T, Void}` would be the preferable solution.  All of this surprised me because it seemed to run counter to much of what I'd heard at JuliaCon 2015, where there seemed to be some emphasis on `Nullable`, `NullableArray`, and the important improvements that would yield for `DataFrames`.

To what degree is type-constrained dispatch optimization being considered for Julia 1.0?  Is it strong enough to advise against using `Nullable`?

nalimilan commented:
Always returning a `Nullable` is an interesting option. Unfortunately, we need to find a deprecation path that doesn't break existing code, so this means we need two release cycles do do it. I think we should go with `tryget` for now, and then, if we decide to always return a `Nullable`, deprecate the current  `get` in 0.6, so that we can rename `tryget` to `get` in 0.7/1.0.

The same strategy could be applied to `tryparse`. This is also the same situation as for `findfirst` (https://github.com/JuliaLang/julia/pull/15755). We need a systematic plan to evaluate whether and how we should make that general shift. Could a JuliaLang owner create a `nullable` tag to allow grouping these issues together?

dbeach24 commented:
@nalimilan  I was *not* suggesting that the 3-arg form of `get` should return a `Nullable` -- only the 2-arg form, where no default is provided.  If you have an acceptable default value, then the `Nullable` really isn't necessary.  In my question above, I was asking if there were plans to optimize `Union` type dispatch that would soon make the `Nullable` type obsolete.  If that is the case, then we should think twice before further embracing `Nullable` in Base.  Sorry for any confusion.

nalimilan commented:
Even if only the two-argument form is changed, this needs a deprecation path to avoid breaking existing code.

dbeach24 commented:
@nalimilan There *is no* existing 2-arg form.  This is the proposed *addition*, not a proposed change.

nalimilan commented:
Sorry, I would have sweared it existed. So it's actually a good thing it wasn't added before, as it makes it much easier to use it instead of `tryget`. +1

dbeach24 commented:
Just updated the PR to implement:

```jl
get(h::Associative, key) -> Nullable
```

Could someone please review?

rfourquet reviewed (approved):
This needs rebasing (e.g. `PriorityQueue`), but looks good to me. I had forgotten about this issue and was about to start a PR doing exactly this. I really don't see what fits better a two-arg `get` than this, and this seems to be exactly the kind of function `Nullable` is designed for (not in the statitistician meaning of course). I'm just uncertain about the deprecation problem If nullable gets renamed to `Option` or some other name. If this happens "soon", could be better to wait for the new name first (and use `tryget` in the meantime). But this functionality is needed, can someone make a decision?

nalimilan commented:
This is the posterchild for the `Union{Some{T}, Null}` type which has been evoked as a replacement for `Nullable` (https://github.com/JuliaLang/julia/issues/22682). But clearly we need to make a decision regarding `Nullable` before fixing this issue.

KristofferC commented:
Nullable is no longer in Base

nalimilan commented:
The issue itself is still relevant though. See PR https://github.com/JuliaLang/julia/pull/25131.
